### PR TITLE
docs: Cloud Hypervisor API documentation

### PR DESCRIPTION
**Pull request by sameo**
We document both the internal and external APIs, and how they relate to
each others.

Fixes: #318

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Cloud Hypervisor API
+
+The Cloud Hypervisor API is made of 2 distinct interfaces:
+
+1. **The external API**. This is the user facing API. Users and operators can 
+   control and manage Cloud Hypervisor through either a REST API or a Command
+   Line Interface (CLI).
+
+2. **The internal API**, based on [Rust's Multi-Producer, Single-Consumer (MSPC)](https://doc.rust-lang.org/std/sync/mpsc/) module.
```
Comment (jiangliu) [inline docs/api.md]: MPSC:)

Comment (sameo) [reply docs/api.md]: I feel watched :-) Thanks @jiangliu, fixed now.

Comment (jiangliu) [reply docs/api.md]: We are redesign the API interface too, why you switch from restful API to non-restful API?

Comment (sameo) [reply docs/api.md]: The external API is still a REST API. We haven't changed anything, we're only documenting the external API and its relationship with the internal one.

**Comment by sameo** [comment]
I'll add one internal API example as the last section of this document, but this is mostly ready to review @sboeuf @rbradford .

File: docs/api.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+```shell
+#!/bin/bash
+
+curl --unix-socket /tmp/cloud-hypervisor.sock -i \
+     -X PUT 'http://localhost/api/v1/vm.create'  \
+     -H 'Accept: application/json'               \
+     -H 'Content-Type: application/json'         \
+     -d '{
+         "cpus":{"boot_vcpus": 4, "max_vcpus": 4},
+         "kernel":{"path":"/opt/kata/vmlinux-virtio-fs-virtio-iommu"},
```
Comment (sboeuf) [inline docs/api.md]: Just a bit weird to have some Kata reference in here.

File: docs/api.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+If you want to inspect the VMM, or control the VM after launching Cloud
+Hypervisor from the CLI, you must use the [REST API](#rest-api).
+
+From the CLI, one can either:
+
+1. Create and boot a complete virtual machine by using the CLI options to build
+   the VM config. Run `cloud-hypervisor --help` for a complete list of CLI
+   options. As soon as the `cloud-hypervisor` binary is launched, the
+   [REST API](#rest-api) is available for controlling and managing the VM.
+1. Only start the [REST API](#rest-api) server when not passing any VM
```
Comment (sboeuf) [inline docs/api.md]: Following the "one can either:" sentence before, I would replace `Only start` with `Start`.

File: docs/api.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+   configuration options. The VM can then asynchronously created and booted
```
Comment (sboeuf) [inline docs/api.md]: s/can then/can then be/

File: docs/api.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+   by sending HTTP commands to the [REST API](#rest-api). Check the
+   [REST API examples](#rest-api-examples) section for more details.
+
+### REST API and CLI Architecural Relationship
```
Comment (sboeuf) [inline docs/api.md]: s/Architecural/Architectural/

File: docs/api.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+Commands are [MPSC](https://doc.rust-lang.org/std/sync/mpsc/) based messages and
+are received and processed by the VMM control loop.
+
+In order for the VMM control loop to respond to any internal API command, it
+must be able to send a response back to the MSPC sender. For that purpose, all
+internal API command payload carry the [Sender](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html)
+end of an [MPSC](https://doc.rust-lang.org/std/sync/mpsc/) channel. 
+
+The sender of any internal API command is thus responsible for:
```
Comment (sboeuf) [inline docs/api.md]: nit: I would replace `thus` with `therefore`

File: docs/api.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+1. Passing the [Sender](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html)
+   end of the response channel as part of the internal API command payload.
+1. Waiting for the internal API command's response on the [Receiver](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html)
+   end of the response channel.
+
+#### Example
+
+In order to illustrate the internal API operations, we are going to look
+at the complete VM creation flow, from the [REST API](#rest-api) call, to
+the reply the external user will receive.
```
Comment (sboeuf) [inline docs/api.md]: Is the example missing?

Comment (sboeuf) [reply docs/api.md]: Nevermind, I just saw your comment:
>I'll add one internal API example as the last section of this document

File: docs/api.md (hunk 3 above)
Comment (likebreath) [reply docs/api.md]: Have the similar feeling, and think of the following alternative phrase:
"Start the [REST API](#rest-api) server only by not passing any VM ..."

**Comment by sameo** [comment]
@sboeuf @likebreath All comments addressed, and I added an end to end example as the last section of the document.

File: docs/api.md [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+```shell
+#!/bin/bash
+
+curl --unix-socket /tmp/cloud-hypervisor.sock -i -X PUT 'http://localhost/api/v1/vm.shutdown'
+```
+
+### Command Line Interface
+
+The Cloud Hypervisor Command Line Interface (CLI) can only be used for launching
+the Cloud Hypevisor binary, i.e. it can not be used for controlling the VMM or
```
Comment (likebreath) [inline docs/api.md]: Typo: ~~Hypevisor~~ Hypervisor

Comment (sameo) [reply docs/api.md]: Fixed, thanks.

File: docs/api.md [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+[CLI](#command-line-interface) interfaces.
+
+The API follows a command-response scheme that closely maps the [REST API](#rest-api).
+Any command must be replied to with a response.
+
+Commands are [MPSC](https://doc.rust-lang.org/std/sync/mpsc/) based messages and
+are received and processed by the VMM control loop.
+
+In order for the VMM control loop to respond to any internal API command, it
+must be able to send a response back to the MSPC sender. For that purpose, all
```
Comment (likebreath) [inline docs/api.md]: Nit: MSPC -> [MPSC](https://doc.rust-lang.org/std/sync/mpsc/)

File: docs/api.md [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+        * [Create a Virtual Machine](#create-a-virtual-machine)
+      - [Boot a Virtual Machine](#boot-a-virtual-machine)
+      - [Dump a Virtual Machine Information](#dump-a-virtual-machine-information)
+      - [Reboot a Virtual Machine](#reboot-a-virtual-machine)
+      - [Shut a Virtual Machine Down](#shut-a-virtual-machine-down)
+    + [Command Line Interface](#command-line-interface)
+    + [REST API and CLI Architecural Relationship](#rest-api-and-cli-architecural-relationship)
+  * [Internal API](#internal-api)
+    + [Goals and Design](#goals-and-design)
+  * [End to End Example](#end-to-end-example)
```
[the remark this answers is not in the document]
Comment (sameo) [reply docs/api.md]: Fixed, thanks.

File: docs/api.md (hunk 9 above)
Comment (sameo) [reply docs/api.md]: Fixed as well, thanks.

File: docs/api.md [hunk 11]
```diff
@@ -0,0 +1,371 @@
+- [Cloud Hypervisor API](#cloud-hypervisor-api)
+  * [External API](#external-api)
+    + [REST API](#rest-api)
+      - [Location and availability](#location-and-availability)
+      - [Endpoints](#endpoints)
+		* [Virtual Machine Manager (VMM) Actions](#virtual-machine-manager--vmm--actions)
```
Comment (likebreath) [inline docs/api.md]: Typo: ~~#virtual-machine-manager--vmm--action~~ - > `#virtual-machine-manager-vmm-actions`

File: docs/api.md [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+		* [Virtual Machine (VM) Actions](#virtual-machine--vm--actions)
```
Comment (likebreath) [inline docs/api.md]: Typo: ~~#virtual-machine-manager--vmm--actions~~ -> `#virtual-machine-vm-actions`

File: docs/api.md [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+      - [REST API Examples](#rest-api-examples)
+        * [Create a Virtual Machine](#create-a-virtual-machine)
+	    * [Boot a Virtual Machine](#boot-a-virtual-machine)
+        * [Dump a Virtual Machine Information](#dump-a-virtual-machine-information)
+        * [Reboot a Virtual Machine](#reboot-a-virtual-machine)
+        * [Shut a Virtual Machine Down](#shut-a-virtual-machine-down)
+    + [Command Line Interface](#command-line-interface)
+    + [REST API and CLI Architecural Relationship](#rest-api-and-cli-architecural-relationship)
```
Comment (likebreath) [inline docs/api.md]: Typo:
~~`[REST API and CLI Architecural Relationship](#rest-api-and-cli-architecural-relationship)`~~
`[REST API and CLI Architectural Relationship](#rest-api-and-cli-architectural-relationship)`

Comment (sameo) [reply docs/api.md]: :-/ Thanks for the thorough review @likebreath 
It should be clean now...

**Pull request merged**